src/modules/IntegrationTracking.jl

```diff
@@ -387,53 +387,74 @@ end
+    # kinetic momenta, not canonical momenta
+    px = v[i,PXI] + (v[i,YI] * kn[1] / 2) * (mm[1] == 0) 
+    py = v[i,PYI] - (v[i,XI] * kn[1] / 2) * (mm[1] == 0)
+  
 
-  rel_p = 1 + v[i,PZI]
-  beta_gamma = rel_p / tilde_m
-  gamma = sqrt(1 + beta_gamma^2)
-  pl2 = rel_p^2 - px^2 - py^2
-  coords.state[i] = ifelse(pl2 <= 0 && coords.state[i] == State.Alive, State.Lost, coords.state[i])
-  alive = ifelse(coords.state[i]==State.Alive, 1, 0) 
-  pl = sqrt(pl2 + (alive-1)*(pl2-1))
+    rel_p = 1 + v[i,PZI]
+    beta_gamma = rel_p / tilde_m
+    gamma = sqrt(1 + beta_gamma^2)
+    pl2 = rel_p^2 - px^2 - py^2
+    coords.state[i] = ifelse(pl2 <= 0 && coords.state[i] == State.Alive, State.Lost, coords.state[i])
+    alive = ifelse(coords.state[i]==State.Alive, 1, 0) 
+    pl = sqrt(pl2 + (alive-1)*(pl2-1))
 
-  beta_hat = SA[px, py, pl] / rel_p
+    beta_hat = SA[px, py, pl] / rel_p
 
-  bx, by = ExactTracking.normalized_field!(mm, kn, ks, v[i,XI], v[i,YI], -1)
-  bz = kn[1] * (mm[1] == 0)
-  b_field = SA[bx, by, bz]
+    bx, by = ExactTracking.normalized_field(mm, kn, ks, v[i,XI], v[i,YI], -1)
+    bz = kn[1] * (mm[1] == 0)
+    b_field = SA[bx, by, bz]
 
-  dot = b_field[1]*beta_hat[1] + b_field[2]*beta_hat[2] + b_field[3]*beta_hat[3]
-  b_para = dot * beta_hat
-  b_perp = b_field - b_para
-  
-  omega = (1 + a*gamma)*b_perp + (1 + a)*b_para
-  omega = -(1 + g*v[i,XI])/pl * omega
-  omega = omega + SA[0, g, 0]
-  omega = alive * omega
+    dot = b_field[1]*beta_hat[1] + b_field[2]*beta_hat[2] + b_field[3]*beta_hat[3]
+    b_para = dot * beta_hat
+    b_perp = b_field - b_para
+
+    coeff1 = -(1 + g*v[i,XI])/pl
+    coeff2 = 1 + a*gamma
+    coeff3 = 1 + a
+
+    ox = coeff1 * (coeff2 * b_perp[1] + coeff3 * b_para[1])
+    oy = coeff1 * (coeff2 * b_perp[2] + coeff3 * b_para[2]) + g
+    oz = coeff1 * (coeff2 * b_perp[3] + coeff3 * b_para[3])
+
+    ox = alive * L * ox
+    oy = alive * L * oy
+    oz = alive * L * oz
+
+    omega = SA[ox, oy, oz]
+  end end
+
+  #omega = (1 + a*gamma) .* b_perp .+ (1 + a) .* b_para
+  #omega = -(1 + g*v[i,XI])/pl .* omega
+  #omega = omega .+ SA[0, g, 0]
+  #omega = L .* omega
+  #omega = alive .* omega
+  # The above lines currently give wrong numbers with FastGTPSA
 
   return omega
 end
 
 
-@makekernel fastgtpsa=false function rotate_spin!(i, coords::Coords, a, g, tilde_m, mm, kn, ks, L)
+@makekernel fastgtpsa=true function rotate_spin!(i, coords::Coords, a, g, tilde_m, mm, kn, ks, L)
   """
   This function rotates particle i's quaternion according to the multipoles present.
   """
   q2 = coords.q
-  q1 = expq(-L/2 * omega(i, coords, a, g, tilde_m, mm, kn, ks))
-  q3 = quat_mul(q1, q2[i,:])
-  q2[i,Q0], q2[i,QX], q2[i,QY], q2[i,QZ] = q3[Q0], q3[QX], q3[QY], q3[QZ]
+  q1 = expq(-omega(i, coords, a, g, tilde_m, mm, kn, ks, L)/2)
+  @inbounds begin
+    q3 = quat_mul(q1, SA[q2[i,Q0], q2[i,QX], q2[i,QY], q2[i,QZ]])
+    q2[i,Q0], q2[i,QX], q2[i,QY], q2[i,QZ] = q3[Q0], q3[QX], q3[QY], q3[QZ]
+  end
```

Comment:
```suggestion
  q3 = quat_mul(q1, SA[q2[i,Q0], q2[i,QX], q2[i,QY], q2[i,QZ]])
  q2[i,Q0], q2[i,QX], q2[i,QY], q2[i,QZ] = q3[Q0], q3[QX], q3[QY], q3[QZ]
```

`@makekernel` includes automatically a `@inbounds`